File: keyway/src/domains/identity/infra/oidc.rs
//! The browser door.
//!
//! Authorization code flow with PKCE against any OIDC issuer. keyway assumes
//! nothing about how an issuer spells things (ADR-0003): the claim carrying
//! groups is named in config, and group names are matched exactly — an issuer
//! that wants a grant to a parent group to cover the teams inside it emits the
//! ancestors in the claim.

use crate::config::Oidc as OidcConfig;
use crate::domains::identity::entity::Role;
use eyre::{Context as _, bail};
use openidconnect::core::{CoreAuthenticationFlow, CoreClient, CoreProviderMetadata};
use openidconnect::{
    AuthorizationCode, ClientId, ClientSecret, CsrfToken, IssuerUrl, Nonce, PkceCodeChallenge,
    PkceCodeVerifier, RedirectUrl, Scope, TokenResponse,
};
use serde_json::Value;

/// A configured issuer, discovered once at boot.
pub struct Oidc {
    client: CoreClient<
        openidconnect::EndpointSet,
        openidconnect::EndpointNotSet,
        openidconnect::EndpointNotSet,
        openidconnect::EndpointNotSet,
        openidconnect::EndpointMaybeSet,
        openidconnect::EndpointMaybeSet,
    >,
    http: openidconnect::reqwest::Client,
    groups_claim: String,
    roles_claim: String,
    role_prefix: String,
}

/// Who signed in, as the claim describes them.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SignedIn {
    pub handle: String,
    pub email: String,
    pub name: String,
    pub groups: Vec<String>,
    pub roles: Vec<Role>,
}

/// What a redirect to the issuer needs remembering across it.
pub struct Pending {
    pub authorize_url: String,
    pub csrf: String,
    pub nonce: String,
    pub pkce_verifier: String,
}

impl Oidc {
    /// Discovers the issuer.
    ///
    /// At boot rather than per request: a console that only discovers its
    /// issuer when somebody tries to sign in is one that looks healthy while
    /// being unusable.
    ///
    /// # Errors
    ///
    /// When discovery fails or the configuration is not a valid issuer.
    pub async fn discover(config: &OidcConfig) -> eyre::Result<Self> {
        let http = openidconnect::reqwest::ClientBuilder::new()
            // An issuer that redirects is an issuer being impersonated.
            .redirect(openidconnect::reqwest::redirect::Policy::none())
            .build()
            .wrap_err("building an http client")?;

        let issuer = IssuerUrl::new(config.issuer.clone()).wrap_err("the oidc issuer url")?;
        let metadata = CoreProviderMetadata::discover_async(issuer, &http)
            .await
            .wrap_err_with(|| format!("discovering {}", config.issuer))?;

        let client = CoreClient::from_provider_metadata(
            metadata,
            ClientId::new(config.client_id.clone()),
            Some(ClientSecret::new(config.client_secret.clone())),
        )
        .set_redirect_uri(
            RedirectUrl::new(config.redirect_url.clone()).wrap_err("the oidc redirect url")?,
        );

        Ok(Self {
            client,
            http,
            groups_claim: config.groups_claim.clone(),
            roles_claim: config.roles_claim.clone(),
            role_prefix: config.role_prefix.clone(),
        })
    }

    /// Where to send somebody, and what to remember while they are gone.
    #[must_use]
    pub fn start(&self) -> Pending {
        let (challenge, verifier) = PkceCodeChallenge::new_random_sha256();
        let (url, csrf, nonce) = self
            .client
            .authorize_url(
                CoreAuthenticationFlow::AuthorizationCode,
                CsrfToken::new_random,
                Nonce::new_random,
            )
            .add_scope(Scope::new("openid".to_owned()))
            .add_scope(Scope::new("profile".to_owned()))
            .add_scope(Scope::new("email".to_owned()))
            .add_scope(Scope::new("groups".to_owned()))
            .set_pkce_challenge(challenge)
            .url();

        Pending {
            authorize_url: url.to_string(),
            csrf: csrf.secret().clone(),
            nonce: nonce.secret().clone(),
            pkce_verifier: verifier.secret().clone(),
        }
    }

    /// Exchanges the code for an identity.
    ///
    /// # Errors
    ///
    /// When the exchange fails, the id token is absent or its claims do not
    /// verify.
    pub async fn finish(
        &self,
        code: &str,
        nonce: &str,
        pkce_verifier: &str,
    ) -> eyre::Result<SignedIn> {
        let tokens = self
            .client
            .exchange_code(AuthorizationCode::new(code.to_owned()))?
            .set_pkce_verifier(PkceCodeVerifier::new(pkce_verifier.to_owned()))
            .request_async(&self.http)
            .await
            .wrap_err("exchanging the authorization code")?;

        let Some(id_token) = tokens.id_token() else {
            bail!("the issuer returned no id token");
        };
        let claims = id_token
            .claims(
                &self.client.id_token_verifier(),
                &Nonce::new(nonce.to_owned()),
            )
            .wrap_err("verifying the id token")?;

        // `preferred_username` is the handle everything keys and logs on; the
        // subject is stable but unreadable, and an audit log full of uuids
        // answers nobody's question.
        let handle = claims.preferred_username().map_or_else(
            || claims.subject().as_str().to_owned(),
            |u| u.as_str().to_owned(),
        );

        let extra = serde_json::to_value(claims).unwrap_or(Value::Null);

        Ok(SignedIn {
            handle,
            email: claims
                .email()
                .map(|e| e.as_str().to_owned())
                .unwrap_or_default(),
            name: claims
                .name()
                .and_then(|n| n.get(None))
                .map(|n| n.as_str().to_owned())
                .unwrap_or_default(),
            groups: strings_at(&extra, &self.groups_claim),
            roles: strings_at(&extra, &self.roles_claim)
                .iter()
                .filter_map(|name| name.strip_prefix(&self.role_prefix).and_then(Role::parse))
                .collect(),
        })
    }
}

/// Reads a claim by dotted path, e.g. `realm_access.roles`.
///
/// A path rather than a key because issuers nest: Keycloak puts realm roles
/// under `realm_access.roles`, and a flat lookup would find nothing and grant
/// nothing, silently.
fn strings_at(claims: &Value, path: &str) -> Vec<String> {
    let mut current = claims;
    for segment in path.split('.') {
        match current.get(segment) {
            Some(next) => current = next,
            None => return Vec::new(),
        }
    }
    match current {
        Value::Array(items) => items
            .iter()
            .filter_map(|v| v.as_str().map(ToOwned::to_owned))
            .collect(),
        // A single string is a claim with one value, which some issuers emit
        // rather than a one-element array.
        Value::String(one) => vec![one.clone()],
        _ => Vec::new(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn a_flat_claim_is_read() {
        let claims = json!({ "groups": ["SRE", "platform"] });
        assert_eq!(strings_at(&claims, "groups"), ["SRE", "platform"]);
    }

    #[test]
    fn a_nested_claim_is_read_by_path() {
        // Keycloak's shape. A flat lookup would find nothing and grant
        // nothing, silently.
        let claims = json!({ "realm_access": { "roles": ["keyway:admin"] } });
        assert_eq!(strings_at(&claims, "realm_access.roles"), ["keyway:admin"]);
    }

    #[test]
    fn a_single_string_claim_is_one_value() {
        let claims = json!({ "groups": "SRE" });
        assert_eq!(strings_at(&claims, "groups"), ["SRE"]);
    }

    #[test]
    fn a_missing_claim_is_empty_rather_than_an_error() {
        // Somebody with no groups is not a failure, and refusing to sign them
        // in would make "no teams yet" indistinguishable from a broken issuer.
        let claims = json!({ "sub": "abc" });
        assert!(strings_at(&claims, "groups").is_empty());
        assert!(strings_at(&claims, "realm_access.roles").is_empty());
    }

    #[test]
    fn a_claim_of_the_wrong_shape_is_empty() {
        let claims = json!({ "groups": { "not": "a list" } });
        assert!(strings_at(&claims, "groups").is_empty());
    }
}

File: internal/identity/infra/oidc.go
// The browser door.
//
// Authorization code flow with PKCE against any OIDC issuer. keyway assumes
// nothing about how an issuer spells things (ADR-0003): the claim carrying
// groups is named in config, and group names are matched exactly — an issuer
// that wants a grant to a parent group to cover the teams inside it emits the
// ancestors in the claim.

package infra

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"strings"

	"github.com/coreos/go-oidc/v3/oidc"
	"golang.org/x/oauth2"

	"github.com/kotsmile/keyway/config"
	"github.com/kotsmile/keyway/internal/identity/entity"
	identityservice "github.com/kotsmile/keyway/internal/identity/service"
)

// Oidc is a configured issuer, discovered once at boot.
//
// It implements identityservice.Issuer, which is what the transport holds:
// Pending and SignedIn are the identity domain's vocabulary, and everything
// about discovery documents, PKCE verifiers and claim paths stops here.
type Oidc struct {
	oauth       oauth2.Config
	verifier    *oidc.IDTokenVerifier
	http        *http.Client
	groupsClaim string
	rolesClaim  string
	rolePrefix  string
}

var _ identityservice.Issuer = (*Oidc)(nil)

// Discover discovers the issuer.
//
// At boot rather than per request: a console that only discovers its issuer
// when somebody tries to sign in is one that looks healthy while being
// unusable.
func Discover(ctx context.Context, cfg config.Oidc) (*Oidc, error) {
	client := &http.Client{
		// An issuer that redirects is an issuer being impersonated.
		CheckRedirect: func(*http.Request, []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	provider, err := oidc.NewProvider(oidc.ClientContext(ctx, client), cfg.Issuer)
	if err != nil {
		return nil, fmt.Errorf("discovering %s: %w", cfg.Issuer, err)
	}

	return &Oidc{
		oauth: oauth2.Config{
			ClientID:     cfg.ClientID,
			ClientSecret: cfg.ClientSecret,
			RedirectURL:  cfg.RedirectURL,
			Endpoint:     provider.Endpoint(),
			Scopes:       []string{oidc.ScopeOpenID, "profile", "email", "groups"},
		},
		verifier:    provider.Verifier(&oidc.Config{ClientID: cfg.ClientID}),
		http:        client,
		groupsClaim: cfg.GroupsClaim,
		rolesClaim:  cfg.RolesClaim,
		rolePrefix:  cfg.RolePrefix,
	}, nil
}

// Start is where to send somebody, and what to remember while they are gone.
func (o *Oidc) Start() identityservice.Pending {
	verifier := oauth2.GenerateVerifier()
	csrf := randomToken()
	nonce := randomToken()
	return identityservice.Pending{
		AuthorizeURL: o.oauth.AuthCodeURL(csrf,
			oauth2.S256ChallengeOption(verifier), oidc.Nonce(nonce)),
		CSRF:         csrf,
		Nonce:        nonce,
		PKCEVerifier: verifier,
	}
}

// Finish exchanges the code for an identity.
//
// It errors when the exchange fails, the id token is absent or its claims do
// not verify.
func (o *Oidc) Finish(
	ctx context.Context, code, nonce, pkceVerifier string,
) (*identityservice.SignedIn, error) {
	ctx = oidc.ClientContext(ctx, o.http)
	tokens, err := o.oauth.Exchange(ctx, code, oauth2.VerifierOption(pkceVerifier))
	if err != nil {
		return nil, fmt.Errorf("exchanging the authorization code: %w", err)
	}

	raw, ok := tokens.Extra("id_token").(string)
	if !ok {
		return nil, errors.New("the issuer returned no id token")
	}
	idToken, err := o.verifier.Verify(ctx, raw)
	if err != nil {
		return nil, fmt.Errorf("verifying the id token: %w", err)
	}
	if idToken.Nonce != nonce {
		return nil, errors.New("verifying the id token: the nonce did not match")
	}

	var claims map[string]any
	if err := idToken.Claims(&claims); err != nil {
		return nil, fmt.Errorf("reading the id token claims: %w", err)
	}

	// `preferred_username` is the handle everything keys and logs on; the
	// subject is stable but unreadable, and an audit log full of uuids
	// answers nobody's question.
	username, _ := claims["preferred_username"].(string)
	handle, handleErr := entity.NewHandle(username)
	if handleErr != nil {
		// A claim with no usable username at all. The subject is always
		// present and always non-empty, so this is the fallback that cannot
		// itself fail.
		handle, handleErr = entity.NewHandle(idToken.Subject)
		if handleErr != nil {
			return nil, fmt.Errorf("reading the id token claims: the issuer named nobody")
		}
	}
	email, _ := claims["email"].(string)
	name, _ := claims["name"].(string)

	// Only prefixed words are candidates: a realm namespaces its keyway roles
	// (`keyway:admin`) precisely so its other systems' roles are not read as
	// keyway's. What survives the prefix is then a role word like any other,
	// and unknown ones are dropped and named — the same accept-and-warn the
	// dev_roles list gets, from the same function.
	var prefixed []string
	for _, word := range stringsAt(claims, o.rolesClaim) {
		if stripped, ok := strings.CutPrefix(word, o.rolePrefix); ok {
			prefixed = append(prefixed, stripped)
		}
	}
	roles, unknown := entity.ParseRoles(prefixed)
	if len(unknown) > 0 {
		slog.Warn("the roles claim carries keyway-prefixed roles this build does not have",
			"user", handle.String(), "unknown", unknown, "known", entity.RoleWords())
	}

	groups, dropped := entity.GroupNamesOf(stringsAt(claims, o.groupsClaim))
	if len(dropped) > 0 {
		slog.Warn("the groups claim carries entries that name no group; they are ignored",
			"user", handle.String(), "dropped", len(dropped))
	}

	return &identityservice.SignedIn{
		Handle: handle,
		Email:  email,
		Name:   name,
		Groups: groups,
		Roles:  roles,
	}, nil
}

// stringsAt reads a claim by dotted path, e.g. `realm_access.roles`.
//
// A path rather than a key because issuers nest: Keycloak puts realm roles
// under `realm_access.roles`, and a flat lookup would find nothing and grant
// nothing, silently.
func stringsAt(claims map[string]any, path string) []string {
	var current any = claims
	for _, segment := range strings.Split(path, ".") {
		object, ok := current.(map[string]any)
		if !ok {
			return nil
		}
		current, ok = object[segment]
		if !ok {
			return nil
		}
	}
	switch value := current.(type) {
	case []any:
		words := make([]string, 0, len(value))
		for _, item := range value {
			if s, ok := item.(string); ok {
				words = append(words, s)
			}
		}
		return words
	case string:
		// A single string is a claim with one value, which some issuers emit
		// rather than a one-element array.
		return []string{value}
	}
	return nil
}

// randomToken is 32 bytes of entropy, url-safe. Used for the csrf state and
// the nonce; crypto/rand never fails on a platform this runs on.
func randomToken() string {
	buf := make([]byte, 32)
	if _, err := rand.Read(buf); err != nil {
		panic(err) // Documented to never happen since Go 1.24.
	}
	return base64.RawURLEncoding.EncodeToString(buf)
}
